apps/hash-frontend/src/components/hooks/block-protocol-functions/knowledge/use-block-protocol-update-entity.ts

```diff
@@ -22,69 +21,67 @@ export const useBlockProtocolUpdateEntity = (
     fetchPolicy: "no-cache",
   });
 
-  // @ts-expect-error todo-0.3 fix mismatch between EntityId in @blockprotocol/graph and HASH
-  const updateEntity: EmbedderGraphMessageCallbacks["updateEntity"] =
-    useCallback(
-      async ({ data }) => {
-        if (readonly) {
-          return {
-            errors: [
-              {
-                code: "FORBIDDEN",
-                message: "Operation can't be carried out in readonly mode",
-              },
-            ],
-          };
-        }
+  const updateEntity = useCallback<UpdateEntityMessageCallback>(
+    async ({ data }) => {
+      if (readonly) {
+        return {
+          errors: [
+            {
+              code: "FORBIDDEN",
+              message: "Operation can't be carried out in readonly mode",
+            },
+          ],
+        };
+      }
+
+      if (!data) {
+        return {
+          errors: [
+            {
+              code: "INVALID_INPUT",
+              message: "'data' must be provided for updateEntity",
+            },
+          ],
+        };
+      }
 
-        if (!data) {
-          return {
-            errors: [
-              {
-                code: "INVALID_INPUT",
-                message: "'data' must be provided for updateEntity",
-              },
-            ],
-          };
-        }
+      const {
+        entityId,
+        entityTypeId,
+        leftToRightOrder,
+        rightToLeftOrder,
+        properties,
+      } = data;
 
-        const {
-          entityId,
+      const { data: updateEntityResponseData } = await updateFn({
+        variables: {
+          entityId, // @todo-0.3 consider validating that this matches the id format,
```

Comment:
I believe it is still relevant. We've typed it within the function but there's nothing stopping a block from passing some garbage string, so we may want to consider some form of validation step here.

This is a wider issue whereby we are nicely typing the message callbacks for convenience, but are implicitly casting a bunch of things away from what the BP spec guarantees. We do this mostly because we also use the hooks all over the frontend _as_ the interface to our backend.

I think a potential solution would involve introducing another layer in the middle. So we have
BP Layer (plain imports) <-Transform A-> Nicely Typed Layer <-Transform B-> GraphQL

Whereby the frontend can call the Nicely Typed Layer but we can add secondary validations and such within Transform A for some proper type safety. (It would also mean we wouldn't have to be redefining all of the message callbacks..